### PR TITLE
Tag based struct validation

### PR DESCRIPTION
**Pull request by parrasajad**
## Proposed changes

* Add validate tag's to structs and add Yaml decode wrapper with built-in validation logic
* Remove individual validate methods in the each of the reporting providers
## Checklist

- [x] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [x] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: v2/internal/runner/options.go [hunk 1]
```diff
@@ -77,6 +79,18 @@ func hasStdin() bool {
 
 // validateOptions validates the configuration options passed
 func validateOptions(options *types.Options) error {
+	validate := validator.New()
+	if err := validate.Struct(options); err != nil {
+
```
Comment (Mzack9999) [inline v2/internal/runner/options.go:84]: is it possible to remove the empty line?

File: v2/pkg/reporting/exporters/es/elasticsearch.go [hunk 2]
```diff
@@ -20,19 +19,19 @@ import (
 // Options contains necessary options required for elasticsearch communicaiton
 type Options struct {
 	// IP for elasticsearch instance
-	IP string `yaml:"ip"`
+	IP string `yaml:"ip"  validate:"required,ip"`
 	// Port is the port of elasticsearch instance
-	Port int `yaml:"port"`
+	Port int `yaml:"port"  validate:"required"`
```
Comment (Mzack9999) [inline v2/pkg/reporting/exporters/es/elasticsearch.go:24]: What about:
```
`validate:"gte=0,lte=65535"`
```

File: v2/pkg/utils/yaml/yaml_decode_wrapper.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+func DecodeAndValidate(r io.Reader, v interface{}) error {
+	if err := yaml.NewDecoder(r).Decode(v); err != nil {
+		return err
+	}
+	if validate == nil {
+		validate = validator.New()
+	}
+
+	if err := validate.Struct(v); err != nil {
+
```
Comment (Mzack9999) [inline v2/pkg/utils/yaml/yaml_decode_wrapper.go:24]: is it possible to remove the empty line?

**Review by Mzack9999 — CHANGES_REQUESTED** [review]

File: v2/pkg/utils/yaml/yaml_decode_wrapper.go (hunk 3 above)
Comment (parrasajad) [reply v2/pkg/utils/yaml/yaml_decode_wrapper.go:24]: sure!

**Review by Mzack9999 — APPROVED** [review]

**Pull request merged**